### ISSUE TITLE
autodownloading from LAN server goes to infinite loop

### ISSUE THREAD

**Issue by illwieckz** [Unvanquished/Unvanquished#767]
Hi, when you try to download to a LAN server with custom content, the client downloads the map then goes to an infinite loop.

For example with this [`map-parpax_d05+unv-a37-illwieckz.pk3` file](http://gg.illwieckz.net/dl/unvanquished/pkg/map-parpax_d05%2bunv-a37-illwieckz.pk3) that just adds a minimap to `parpax` map.

When the client connect itself, the LAN server says:

```
Client 0 connecting with 0 challenge ping 
UnnamedPlayer#1 connected 
clientDownload: 13 : beginning "map-parpax_d05+unv-a37-illwieckz_9b920f8e" 
'UnnamedPlayer#1' downloading at sv_dl_maxrate (42000) 
clientDownload: 13 : file "map-parpax_d05+unv-a37-illwieckz_9b920f8e" completed 
[PM_LandJetpack] Landing ignored (hit surface at 9°) 
clientDownload: 13 : beginning "map-parpax_d05+unv-a37-illwieckz_9b920f8e" 
'UnnamedPlayer#1' downloading at sv_dl_maxrate (42000) 
clientDownload: 13 : file "map-parpax_d05+unv-a37-illwieckz_9b920f8e" completed 
[PM_LandJetpack] Landing ignored (hit surface at 2°) 
clientDownload: 13 : beginning "map-parpax_d05+unv-a37-illwieckz_9b920f8e" 
'UnnamedPlayer#1' downloading at sv_dl_maxrate (42000) 
clientDownload: 13 : file "map-parpax_d05+unv-a37-illwieckz_9b920f8e" completed 
[PM_LandJetpack] Landing ignored (hit surface at 32°) 
[PM_LandJetpack] Landing ignored (hit surface at 44°) 
clientDownload: 13 : beginning "map-parpax_d05+unv-a37-illwieckz_9b920f8e" 
'UnnamedPlayer#1' downloading at sv_dl_maxrate (42000) 
clientDownload: 13 : file "map-parpax_d05+unv-a37-illwieckz_9b920f8e" completed 
[PM_LandJetpack] Landing ignored (hit surface at 5°) 
clientDownload: 13 : beginning "map-parpax_d05+unv-a37-illwieckz_9b920f8e" 
'UnnamedPlayer#1' downloading at sv_dl_maxrate (42000) 
clientDownload: 13 : file "map-parpax_d05+unv-a37-illwieckz_9b920f8e" completed 
clientDownload: 13 : beginning "map-parpax_d05+unv-a37-illwieckz_9b920f8e" 
'UnnamedPlayer#1' downloading at sv_dl_maxrate (42000) 
clientDownload: 13 : file "map-parpax_d05+unv-a37-illwieckz_9b920f8e" completed 
clientDownload: 13 : beginning "map-parpax_d05+unv-a37-illwieckz_9b920f8e" 
'UnnamedPlayer#1' downloading at sv_dl_maxrate (42000) 
[PM_LandJetpack] Landing ignored (hit surface at 22°) 
clientDownload: 13 : file "map-parpax_d05+unv-a37-illwieckz_9b920f8e" completed 
clientDownload: 13 : beginning "map-parpax_d05+unv-a37-illwieckz_9b920f8e" 
…
```

And while this infinite loop is running, there is these files in client's `pkg/` dir:

```
map-parpax_d05+unv-a37-illwieckz_9b920f8e.pk3 
map-parpax_d05+unv-a37-illwieckz_9b920f8e.pk3.tmp 
map-parpax_d05+unv-a37-illwieckz.pk3
```

The `map-parpax_d05+unv-a37-illwieckz.pk3` file is fine in the `pkg/` client's directory, if I disconnect and rejoin, it works.

**Comment by illwieckz** [comment]
I just want to add that when someone is stuck in map loading, there is no interface to disconnect or to quit.

While the pk3 is downloading, it's possible to open the console, but the console close itself when the download is complete, so unvanquished try to download a very little file, you just don't have the time to open the console and type `quit`.

So, some users can be very confused.

**Comment by illwieckz** [comment]
perhaps related to #722 and #482?

**Comment by DolceTriade** [comment]
Same as #482, I think.

**Comment by illwieckz** [comment]
Easy way to reproduce (works every time)

1. get an extra package, take one small so you will have fast loops, like the [parpax minimap](http://gg.illwieckz.net/dl/unvanquished/pkg/map-parpax_d05%2bunv-a37-illwieckz.pk3) or the [redscreen bug hotfix](http://gg.illwieckz.net/dl/unvanquished/pkg/unvanquished_0.42.0%2bgit7ae3ea4%2b1.pk3).
2. host a map on LAN
3. start another instance (use another homepath) and connect to server.

So, to summarize:

```sh
# Terminal 1
( cd ~/.unvanquished/pkg/; wget http://gg.illwieckz.net/dl/unvanquished/pkg/map-parpax_d05%2bunv-a37-illwieckz.pk3 )
daemon +devmap parpax
# Terminal 2
daemon -homepath /tmp/plop +connect 127.0.0.1:27960
```

Enjoy.

**Comment by illwieckz** [comment]
Note that the LAN loop only use the internal download mechanism, but with online download loops it goes infinite with http download too, but since http downloads depend on the internal download code, it's probably due to that internal download code.

**Comment by illwieckz** [comment]
`<Ishq> This might sound stupid, but could it possibly be because its not updating the paks after each download?`
`<illwieckz> yes`
`<illwieckz> while the server said looping "clientDownload" `
`<illwieckz> the client said looping "Loading pak"`
`<illwieckz> server side:` http://pastebin.com/dxHHXFCA
`<illwieckz> client side:` http://pastebin.com/3mugiE87

**Comment by Kangz** [comment]
I cannot repro, either in Debug or Release :/

If that's ok, I'll add a bunch of logging to the engine and ask you to repro with the logging enabled so we can gather more info

**Comment by Kangz** [comment]
Also, what is your distribution?

**Comment by illwieckz** [comment]
Wow, perturbed, you and `Ishq can't repro but I can repro on LAN everytime and online sometime (the first time I get this bug was the first time I started Unvanquished 0.41.0 right after the download), what is happening there? Also, many newcomers get it too…

Having more debug print seems needed yes. What can I do more to track down it?

I use Ubuntu 15.04.

**Comment by Kangz** [comment]
I added the debug output and tried on an Ubuntu variant I have without success. Can you repro on LAN with both logs.logLevel.client.pakDownload debug and logs.logLevel.server.clientCommands debug and post the logs? It will give us much more information as to where things are going wrong.

**Comment by illwieckz** [comment]
Yes, I will try it when I will be at home. :-)

**Comment by Kangz** [comment]
Great! Thank you.

**Comment by illwieckz** [comment]
Download loop server-side: http://pastebin.com/js8Dn3fG
Download loop client-side: http://pastebin.com/Uqx121TV

Note: it's easier to get the download loop if `homepath/pkg` does not exist at daemon startup.

**Comment by illwieckz** [comment]
Another strange behavior, connecting from `daemon` from master to my online 0.42.0 s`daemonded`:
http://pastebin.com/JCfj2GEa

The client falls in download loop then disconnect itself from server with a _"Server sending download, but no download was requested"_ error.

**Comment by Kangz** [comment]
[27|23:26:46] <+kangz> The problem is that pak paths can be discarded at startup and never tried again
[27|23:27:09] <+kangz> instead FS::Initialize should store the paths asked somewhere and retry them on FS::RefreshPaks

**Comment by illwieckz** [comment]
Fixed since 0.43 release by 580fbb9 `“Create the homepath at FS initialization and check that homepath & libpath are valid”`

See also 9b49e5a `“Eat the download data even if we didn't expect one”` for a fixed related bug revealed by this one.

The bug was: the game was not looking at `pkg` directory after autodownload if `pkg` was not there at game startup (virgin profile) because the engine discarded this path at startup and was never looking for paks in new-created `pkg` directory at pak load after an autodownload.

The fix by @Amanieu creates the missing directory at startup even if there is nothing there yet, so the engine always have it before and after the first autodownloaded package and virgin profiles are safe since the first unvanquished run.

The other fix by @kangz is to discard unattended data received (probably due to some race condition with udp/http autodownload and when a download loop with very small files attempt (very fast looping).
